odxtools/encodestate.py

```diff
@@ -86,66 +87,145 @@ def emplace_atomic_value(
+                text_encoding = "utf-16-be" if is_highlow_byte_order else "utf-16-le"
+                raw_value = internal_value.encode(text_encoding)
+            elif base_type_encoding == Encoding.ISO_8859_1 or (
+                    base_data_type == DataType.A_ASCIISTRING and base_type_encoding is None):
+                raw_value = internal_value.encode("iso-8859-1")
+            elif base_type_encoding == Encoding.ISO_8859_2:
+                raw_value = internal_value.encode("iso-8859-2")
+            elif base_type_encoding == Encoding.WINDOWS_1252:
+                raw_value = internal_value.encode("cp1252")
+            else:
+                odxraise(f"Specified illegal encoding {base_type_encoding} for string object")
+                raw_value = internal_value.encode("iso-8859-1")
 
             if 8 * len(raw_value) > bit_length:
-                raise EncodeError(f"The string {repr(internal_value)} is too large."
-                                  f" The maximum number of bytes is {bit_length//8}.")
+                odxraise(
+                    f"The internal representation {raw_value!r} is too large "
+                    f"to be encoded. The maximum number of bytes is {(bit_length + 7)//8}.",
+                    EncodeError)
 
-        elif base_data_type == DataType.A_UNICODE2STRING:
-            if not isinstance(internal_value, str):
-                odxraise()
+        # ... signed integers, ...
+        elif base_data_type == DataType.A_INT32:
+            if not isinstance(internal_value, int):
+                odxraise(
+                    f"Internal value must be of integer type, not {type(internal_value).__name__}",
+                    EncodeError)
+
+            if base_type_encoding == Encoding.ONEC:
+                # one-complement
+                if internal_value >= 0:
+                    raw_value = internal_value
+                else:
+                    mask = (1 << bit_length) - 1
+                    raw_value = mask + internal_value
+            elif base_type_encoding in (None, Encoding.TWOC):
+                # two-complement
+                if internal_value >= 0:
+                    raw_value = internal_value
+                else:
+                    mask = (1 << bit_length) - 1
+                    raw_value = mask + internal_value + 1
+            elif base_type_encoding == Encoding.SM:
+                # sign-magnitude
+                if internal_value >= 0:
+                    raw_value = internal_value
+                else:
+                    raw_value = (1 << (bit_length - 1)) + abs(internal_value)
+            else:
+                odxraise(f"Illegal encoding ({base_type_encoding}) specified for "
+                         f"{base_data_type.value}")
+
+                internal_value = raw_value
+
+        # ... unsigned integers, ...
+        elif base_data_type == DataType.A_UINT32:
+            if not isinstance(internal_value, int) or internal_value < 0:
```

Comment:
Didn't you put `u` in bitstruct logic, so that would not happen?

Reply:
this is a user-specified value. for `.decode()` this is correct...

Reply:
User specify physical value, not internal one

Reply:
yes, but e.g., if the compu method is IDENTICAL and the user specifies a negative value, it still arrives here...